CARDS-2199: As a clinician, I can see on the DATA-PRO clinician dashboard whether surveys have been emailed to a patient prior to their appointment

marta- opened the pull request:
- [x] CARDS-2227: PROMs-specific Survey events questionnaire
- [x] CARDS-2201: Augment the Visit information json with the timestamps when notification emails were sent from Survey events
- [x] CARDS-2202: Update the Appointments dashboard widget to display whether survey links had been sent to the patient
- [x] CARDS-2200: Enable the Survey Tracker in proms

To test proms:
1. **on branch `CARDS-2200-test`** build with `mvn clean install -Pdocker`
2. in `compose-cluster/mssql` run `python3 generate_test_proms_sql.py --time_spread_minutes 4320 -n 30 proms_sample.sql`, check that there are events for tomorrow and 3 days from now, and if not, rerun the script
3. in `compose-cluster` run `python3 generate_compose_yaml.py --mssql --cards_project cards4proms --oak_filesystem --dev_docker_image --smtps --smtps_test_container --smtps_test_mail_path ~/cardsmail --composum --adminer --clarity`
4. edit `docker-compose.yml` and add `NIGHTLY_NOTIFICATIONS_SCHEDULE=0 * * * * ? *` to the cardsinitial environment to have emails sent every minute
5. in `compose-cluster` start cards with `docker-compose up`
6. at `http://localhost:1435/?mssql=mssql&username=sa&db=master&ns=path&import=` import the generated `proms_sample.sql` file
7. open `http://localhost:8080/Subjects.importClarity`
8. check that the patients have been imported correctly
9. check that Survey Events have been created
10. wait for one minute, check that patients with a visit 3 days from today have their Initial Notification question answered, patients with a visit tomorrow have their Reminder Notification question answered
11. log in, fill in and submit one of the visits, make sure that the Survey Events has an answer for the Responses Received question
12. go to the clinic dashboard, make sure that the upcoming visits shows the correct email type and date for the patients that were supposed to receive emails

To test regressions in prems:
  
1. **on branch `CARDS-2199`** build with `mvn clean install -Pdocker`
2. in `compose-cluster/mssql` run `python3 ./generate_test_YE_sql.py -n 100 --basedate 2023-05-03 prems_sample.sql` (use 6 days ago as the date)
3. in `compose-cluster` run `python3 generate_compose_yaml.py --mssql --cards_project cards4prems --oak_filesystem --dev_docker_image --smtps --smtps_test_container --smtps_test_mail_path ~/cardsmail --composum --adminer --clarity`
4. edit `docker-compose.yml` and add `NIGHTLY_NOTIFICATIONS_SCHEDULE=0 * * * * ? *` to the cardsinitial environment to have emails sent every minute
5. in `compose-cluster` start cards with `docker-compose up`
6. at `http://localhost:1435/?mssql=mssql&username=sa&db=master&ns=path&import=` import the generated `prems_sample.sql` file
7. open `http://localhost:8080/Subjects.importClarity`
8. check that the patients have been imported correctly
9. check that Survey Events have been created
10. wait for one minute, check that patients with a visit 7 days ago have their Initial Notification question answered, patients with a visit 9 days ago have their Reminder1 Notification question answered, and 11 days ago Reminder 2 Notification
11. generate a token, log in, fill in and submit one of the visits, make sure that the Survey Events has an answer for the Responses Received question

marta- committed:
CARDS-2199 / CARDS-2201: Augment the Visit information json with the timestamps when notification emails were sent from Survey events

marta- committed:
CARDS-2199 / CARDS-2202: Update the Appointments dashboard widget to display whether survey links had been sent to the patient

marta- committed:
CARDS-2199 / CARDS-2227: PROMs-specific Survey events questionnaire

sdumitriu committed:
CARDS-2200: Enable the Survey Tracker in proms
Make the SurveyTracker class easy to customize

sdumitriu committed:
[misc] Refactoring to remove code duplication and fix a too-specific method name

sdumitriu committed:
CARDS-2200: Enable the Survey Tracker in proms
Move the class from prems to patient-portal since it is generic enough

sdumitriu committed:
CARDS-2200: Enable the Survey Tracker in proms
Add configuration to enable the survey tracker both in prems and proms

marta- reviewed (commented):
Survey tracker changes look good. Did not test.

sashaandjic commented:
PROMS - ALL GOOD!

I'm having problems when testing regressions in prems:
There is no generate_test_YE_sql.py, I only see generate_test_sql.py.
Also there is a typo in ( proms instead of prems) : 
`python3 generate_compose_yaml.py --mssql --cards_project cards4proms --oak_filesystem --dev_docker_image --smtps --smtps_test_container --smtps_test_mail_path ~/cardsmail --composum --adminer --clarity`

by using `generate_test_sql.py` and with the corrected typo, I am not getting anything imported in prems

sdumitriu commented:
Indeed, I'll update the instructions, they're still modeled on the 1886 branch.

sashaandjic commented:
Small bug: in `prems` the second reminder isn't shown on the dashboard
![image](https://github.com/data-team-uhn/cards/assets/23243244/b6583caf-6549-46c7-85c9-5c9ff8c4d878)

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/config/CopyAnswers/Questionnaires/Visit information.json:
```diff
@@ -3,6 +3,9 @@
   "jcr:reference:surveys_submitted": "/Questionnaires/Visit information/surveys_submitted",
+  "jcr:reference:invitation_sent": "/Questionnaires/Survey events/invitation_sent",
+  "jcr:reference:reminder1_sent": "/Questionnaires/Survey events/reminder1_sent",
+  "jcr:reference:remindoe2_sent": "/Questionnaires/Survey events/reminder2_sent",
```
Typo...
```suggestion
  "jcr:reference:reminder2_sent": "/Questionnaires/Survey events/reminder2_sent",
```

marta- committed:
CARDS-2199: As a clinician, I can see on the DATA-PRO clinician dashboard whether surveys have been emailed to a patient prior to their appointment

Typo fix in CopyAnswers. Typo caused the second reminder timestamp to never be displayed on the clinic dashboard.

marta- commented:
> Small bug: in `prems` the second reminder isn't shown on the dashboard 

Fixed in e36f6de

acrowthe reviewed (approved):
Reviewed, not tested
